Yardoc

durandom opened the pull request:
This is an initial stab at putting the scarce but existent documentation in the source code into a browsable format. I think having something thats actually visible and accessible encourages people to read and _write_ docs.

I've opted for http://yardoc.org/ because I know it.

To add files to the documentation, simply add them to `.yardopts`
To build the documentation run `bundle exec yard` and point your w3c to `doc/index.html`

Maybe we can run this periodically and publish it somewhere.

If this is something we want to do, I would continue with cleaning up docs those 2 files and scan for additional stuff we want to include

@miq-bot add_label documentation

cc @hayesr @Fryguy @chessbyte @blomquisg

Fryguy commented:
yardoc is the format we've been using for documentation elsewhere in the product and in our gems, so that's a good choice.

Fryguy commented on app/models/manager_refresh/inventory_collection.rb:
```diff
@@ -1,3 +1,50 @@
+# Example 1, storing Vm model data into the DB:
+####################################################################################################################
+#   ################################################################################################################
+#   # Example 1.1 Starting with no vms, lets add vm1 and vm2
```
If you're using yarddoc, prefer the `@example` tag here.

chrisarcand replied:
Meh, I'd go through and just convert these two docs to yarddoc, as a good example of a before and after :)

durandom commented on app/models/manager_refresh/inventory_collection.rb:
```diff
@@ -1,3 +1,50 @@
+# Example 1, storing Vm model data into the DB:
+####################################################################################################################
+#   ################################################################################################################
+#   # Example 1.1 Starting with no vms, lets add vm1 and vm2
```
sure - I havent touched the docs at all, just moved it, to produce at least _some_ output.
First wanted to make sure we want to go in that direction...

durandom commented:
@chessbyte @blomquisg @Fryguy so is this something we want to do? Should I work with @hayesr to get this published online?

blomquisg commented:
@durandom can you include some example output?

durandom commented:
@blomquisg the output will be basically the same as http://www.rubydoc.info/github/manageiq/manageiq - but restricted to the files listed in [.yardopts](https://github.com/ManageIQ/manageiq/pull/14475/files#diff-5808b886486adcc2f1820a316cd5652d)

chessbyte commented:
@durandom @blomquisg @Fryguy bump

durandom commented:
I'm still 👍 on exposing the generated docs somewhere on our website.
Should not be too much setup to generate them during the website build process...

chrisarcand commented on .yardopts:
```diff
@@ -0,0 +1,2 @@
+app/models/mixins/supports_feature_mixin.rb
+app/models/manager_refresh/inventory_collection.rb
```
> To add files to the documentation, simply add them to .yardopts

Is this actually what this does, excluding the rest of the files from the documentation? I thought this file specification was for adding files that wouldn't be used by default (`.rb` is a default so I would assume all the docs would be compiled as they currently are)

Also, if this really does just include the files above and nothing else....why? I see no reason you shouldn't just include everything, it's still useful to see a class interface and if anything motivates to start writing docs for things you look up but find there's no documentation written for.

durandom replied:
It will only produce output for the selected files, yes.
Why? Because I rather have few discoverable documentation than a huge ocean of just method names without doc value.

Go to http://www.rubydoc.info/github/manageiq/manageiq - browse a bit around. I guess you wont come back again, cause there is no real documentation in that. Compare that to http://www.rubydoc.info/github/rails/rails/

chrisarcand replied:
I guess ¯\\\_(ツ)_/¯ I think it's still useful to be able to search classes/methods to see if they even exist. Rails isn't 100% fully documented either, it just has a lot more than us :)

Maybe we can leave this as you have it and it can be changed once we have more documentation at all.

chrisarcand commented:
> I'm still 👍 on exposing the generated docs somewhere on our website.

I like that idea; we can do that separately.

mkanoor commented:
There is also swagger_yard https://github.com/livingsocial/swagger_yard which can be used to convert REST API into swagger/OpenAPI specs and auto generate clients in many different languages.

swagger_yard => Swagger (OpenAPI) spec => API clients in multiple languages.

abellotti commented:
👍  on swagger and auto-gen clients in tons of languages.

durandom commented:
@chrisarcand please have a look again.
I reformatted the manager refresh docs and included all manager_refresh in .yardopts

After that we should have a more structured approach towards adding docs to that file.
E.g. picking whats already documented (`git grep '.*#'`) and asking around in the external provider teams what needs to be documented

chessbyte commented:
❤️  that this is back on the radar!

chrisarcand commented on app/models/manager_refresh/inventory_collection.rb:
```diff
@@ -183,7 +182,8 @@ class InventoryCollection
     #            })
     #
-    #          And the labmda is defined as:
+    #        And the labmda is defined as:
```
Typo

chrisarcand commented on app/models/manager_refresh/inventory_collection.rb:
```diff
@@ -151,7 +149,8 @@ class InventoryCollection
     #            })
     #
-    #          And the labmda is defined as:
+    #        And the labmda is defined as:
```
Typo

NickLaMuro commented:
> @durandom You should also add info (as you did in your description) on how to build the documentation locally to the guides.

@chrisarcand @durandom For what it is worth, I personally was mucking around adding some docs to https://github.com/ManageIQ/manageiq/pull/15989, and found that `bundle exec yard server -r` is pretty convenient for testing and viewing generated doc changes, and even subsequent reloads is relatively quick even without the `.yardopts` file to limit what is used.

Thought it was worth mentioning.

durandom commented:
@chrisarcand 

>Here's a perfect example of when I would have changed it and merged it if the Allow maintainers write access checkbox were selected 😄

It's checked for me 😛 But I'm happy to change it tomorrow...

durandom commented:
@NickLaMuro thanks for the hint - I actually used that command as well - will include it in the docs for docs.
@chrisarcand addressed your comments - please have a look again

durandom commented:
https://github.com/ManageIQ/guides/pull/250 is the guides PR

chrisarcand commented:
Let's start here! LGTM
